ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -423,119 +435,253 @@ private void checkOpen() {
+   * @throws JSONException If parsing account data in json fails.
+   */
+  AccountInfoMap(HelixAccountService accountService, Map<String, String> accountMap) throws JSONException {
+    this.accountService = accountService;
+    nameToAccountMap = new HashMap<>();
+    idToAccountMap = new HashMap<>();
+    for (Map.Entry<String, String> entry : accountMap.entrySet()) {
+      String idKey = entry.getKey();
+      String valueString = entry.getValue();
+      Account account;
+      JSONObject accountJson = new JSONObject(valueString);
+      if (idKey == null) {
+        accountService.accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException(
+            "Invalid account record when reading accountMap in ZNRecord because idKey=null");
+      }
+      account = Account.fromJson(accountJson);
+      if (account.getId() != Short.valueOf(idKey)) {
+        accountService.accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException(
+            "Invalid account record when reading accountMap in ZNRecord because idKey and accountId do not match. idKey="
+                + idKey + " accountId=" + account.getId());
+      }
+      if (idToAccountMap.containsKey(account.getId()) || nameToAccountMap.containsKey(account.getName())) {
+        throw new IllegalStateException(
+            "Duplicate account id or name exists. id=" + account.getId() + " name=" + account.getName());
+      }
+      idToAccountMap.put(account.getId(), account);
+      nameToAccountMap.put(account.getName(), account);
     }
+  }
 
-    /**
-     * Gets all the {@link Account}s in this {@code AccountInfoMap} in a {@link Collection}.
-     * @return A {@link Collection} of all the {@link Account}s in this map.
-     */
-    private Collection<Account> getAccounts() {
-      return Collections.unmodifiableCollection(idToAccountMap.values());
+  /**
+   * Gets {@link Account} by its id.
+   * @param id The id to get the {@link Account}.
+   * @return The {@link Account} with the given id, or {@code null} if such an {@link Account} does not exist.
+   */
+  Account getAccountById(Short id) {
+    return idToAccountMap.get(id);
+  }
+
+  /**
+   * Gets {@link Account} by its name.
+   * @param name The id to get the {@link Account}.
+   * @return The {@link Account} with the given name, or {@code null} if such an {@link Account} does not exist.
+   */
+  Account getAccountByName(String name) {
+    return nameToAccountMap.get(name);
+  }
+
+  /**
+   * Gets all the {@link Account}s in this {@code AccountInfoMap} in a {@link Collection}.
+   * @return A {@link Collection} of all the {@link Account}s in this map.
+   */
+  Collection<Account> getAccounts() {
+    return Collections.unmodifiableCollection(idToAccountMap.values());
+  }
+}
+
+/**
+ * An AccountMetadataStore is a storage to keep and retrieve {@link Account} metadata.
+ */
+interface AccountMetadataStore {
```

Comment:
Please refactor this class as it has gotten way too complex.  You have an interface and multiple implementations of it all in this class.

Reply:
It's actually not in the class, but in the same file. I separated the interface and two implementations out of the class but didn't create a new file for each of them, thinking that the legacy one would be removed later any way.
Should I just create separated files?

Reply:
Yeah, probably better to just create separate files in the same package for each.

Reply:
updated